aspell: add all missing official dictionaries

liff opened the pull request:
From:

  ftp://ftp.gnu.org/gnu/aspell/dict/0index.html

Also sets the default homepage of the dictionaries to
the above URL.

The Marathi and Hindi dictionaries both contain `u-deva.{cmap,cset}` files, which cause a conflict. This patch adds those files from the Marathi dictionary into the main aspell package like [Debian does](https://sources.debian.org/src/aspell/0.60.7%7E20110707-5/debian/patches/02_u-deva_common_data.diff).

###### Motivation for this change

I intended to add the Finnish dictionary to `aspellDicts` but noticed there’s a fairly large bunch of official ones not included so why not add them all in one go?

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

liff committed:
aspell: add all missing official dictionaries

From:

  https://ftp.gnu.org/gnu/aspell/dict/0index.html

Use a specialized function, buildOfficialDict to specialize builds of
official aspell dictionaries and add an update script.

liff commented:
Running `nox-review wip` results in a build failure in [`haskellPackages.djinn`](https://hackage.haskell.org/package/djinn) but I’m fairly certain it’s not related.

FRidh commented on pkgs/development/libraries/aspell/default.nix:
```diff
@@ -2,6 +2,17 @@
+
+  # Source for u-deva.cmap and u-deva.cset: use the Marathi
+  # dictionary like Debian does.
+  devaMapsSource = fetchTarball {
```
fetchurl

liff replied:
`fetchurl` gives me the plain tarball file but in this case I just need two files from the extracted tarball.

I could use `fetchurl` and extract the files in manually `postInstall`, though, if that’s preferrable?

FRidh replied:
Ah, I see. No, this is fine.

FRidh replied:
Actually, I am pretty sure this is going to fail on Hydra because it runs in restricted-eval mode. Therefore, considering you wanted to have it unpacked, you could use `fetchzip`. It doesn't matter if the archive is not a zip file.

liff replied:
Ok, thanks, I’ll try it out.

liff replied:
Changed.

FRidh commented:
It's good to add dictionaries for other languages, however, we need a way to maintain these expressions. Basically, we need a script that can generate the data set.

liff commented:
I’ll see if I can come up with a robust script for keeping these up to date.

liff committed:
aspell: remove override of name in buildOfficialDict

The name should include the version number and this is handled just fine by
buildDict.

liff committed:
aspell: use fetchzip instead of fetchTarball to fetch u-deva files

Ekleog reviewed (changes requested):
@liff This looks good, thanks! Just one point: here, it looks like a MitM to someone who runs the update script can execute arbitrary code on the computer running the update script (via the `$versions` substitution), so it's I think a bit too permissive.

Maybe it'd be possible to write it in a more defensive way, to prevent that from happening?

liff commented:
Would it be sufficient to restrict the version string to `[A-Za-z0-9+.-]` instead of `.+` to keep the generated version entries within quotes?

Ekleog commented:
It should do it indeed!

Now, I've just noticed another issue: the `nix` executable interface is not stable yet, so we're not allowed to rely on it yet in `nixpkgs`.

liff committed:
Address review comments

 - Use HTTPS and tighter version regexp to guard against MitM
 - Use nix-instantiate instead of experimental nix eval
 - Handle dictionary-specific meta overrides

liff commented:
Thanks for the review.

Changed from FTP to HTTPS since apparently they are [going to disable FTP at some unspecified time anyway](https://ftp.gnu.org/README), adjusted the regexp, changed from `nix` to `nix-instantiate` and fixed the `meta` thing.

Ekleog commented:
@GrahamcOfBorg eval
@GrahamcOfBorg build aspellDicts.bn

oxij commented:
lynx instead of curl? This is weird.

Ekleog commented:
@liff Do you think you could replace your usage of `lynx` with `curl` and some ad-hoc matching on html elements? If you don't, as it's for an update script it's likely not that bad, but it'd for sure be cleaner to.

7c6f434c commented:
In practice hacks like those in my updater `nixpkgs/pkgs/build-support/upstream-updater/urls-from-page.sh` are almost always sufficient; not sure if I would call them cleaner than actually using an HTML parser.

liff committed:
Scrape directory index with curl instead of lynx

liff commented:
Changed to `curl`.

liff committed:
Fix nix-shell packages for lynx->curl

Ekleog commented:
@GrahamcOfBorg eval
@GrahamcOfBorg build aspellDicts.bn

Ekleog commented:
Looks like everything is good! Thank you, @liff :) Merged!

Ekleog commented:
… And noticed just a bit too late to stop the push: The last 3 commits' messages are not following nixpkgs' convention (while the first 3 were). Mea culpa for not having noticed this.

liff commented:
Oops, sorry. Seems I forgot about that half way when working on the requested changes.
